lib/spack/spack/binary_distribution.py

```diff
@@ -2014,58 +2020,35 @@ def install_single_spec(spec, allow_root=False, unsigned=False, force=False):
     """
-    specfile_name = tarball_name(spec, ".spec.json")
-    signed_specfile_name = tarball_name(spec, ".spec.json.sig")
-    specfile_is_signed = False
+    files = [tarball_name(spec, ".spec.json"), tarball_name(spec, ".spec.json.sig")]
     found_specs = []
 
-    for mirror in spack.mirror.MirrorCollection(mirrors=mirrors).values():
-        buildcache_fetch_url_json = url_util.join(
-            mirror.fetch_url, _build_cache_relative_path, specfile_name
-        )
-        buildcache_fetch_url_signed_json = url_util.join(
-            mirror.fetch_url, _build_cache_relative_path, signed_specfile_name
-        )
-        try:
-            _, _, fs = web_util.read_from_url(buildcache_fetch_url_signed_json)
-            specfile_is_signed = True
-        except (URLError, web_util.SpackWebError, HTTPError) as url_err:
+    for file in files:
+        for mirror in spack.mirror.MirrorCollection(mirrors=mirrors).values():
             try:
-                _, _, fs = web_util.read_from_url(buildcache_fetch_url_json)
-            except (URLError, web_util.SpackWebError, HTTPError) as url_err_x:
-                tty.debug(
-                    "Did not find {0} on {1}".format(
-                        specfile_name, buildcache_fetch_url_signed_json
-                    ),
-                    url_err,
-                    level=2,
-                )
-                tty.debug(
-                    "Did not find {0} on {1}".format(specfile_name, buildcache_fetch_url_json),
-                    url_err_x,
-                    level=2,
+                _, _, fs = web_util.read_from_url(
+                    url_util.join(mirror.fetch_url, _build_cache_relative_path, file)
                 )
+            except (URLError, web_util.SpackWebError, HTTPError) as url_err:
+                tty.debug(url_err, level=2)
                 continue
-        specfile_contents = codecs.getreader("utf-8")(fs).read()
-
-        # read the spec from the build cache file. All specs in build caches
-        # are concrete (as they are built) so we need to mark this spec
-        # concrete on read-in.
-        if specfile_is_signed:
-            specfile_json = Spec.extract_json_from_clearsig(specfile_contents)
-            fetched_spec = Spec.from_dict(specfile_json)
-        else:
-            fetched_spec = Spec.from_json(specfile_contents)
-        fetched_spec._mark_concrete()
 
-        found_specs.append(
-            {
-                "mirror_url": mirror.fetch_url,
-                "spec": fetched_spec,
-            }
-        )
+            utf8_stream = codecs.getreader("utf-8")(fs)
+            fetched_spec = Spec.from_dict(load_possibly_clearsigned_json(utf8_stream))
+
+            # read the spec from the build cache file. All specs in build caches
+            # are concrete (as they are built) so we need to mark this spec
+            # concrete on read-in.
+            fetched_spec._mark_concrete()
+
+            return [
```

Comment:
I'm very slow to understand the logic, but do I get this right: the original idea of `get_mirrors_for_spec` is to do an _offline_ search (local version of remote database), so that remote mirrors are ordered in such a way that it's likely to get a direct hit?

It seems however that with Jonathon's changes this search is no longer offline, since a new copy of the index gets downloaded if its not found locally.

If this is the original idea, I'm happy with that, but it would be great if querying the local/offline cache caused a remote query on failure.

Reply:
> the original idea of get_mirrors_for_spec is to do an offline search (local version of remote database), so that remote mirrors are ordered in such a way that it's likely to get a direct hit?

Yes, that was part of the idea.  Also, during gitlab ci pipeline generation, it gave us the option (`--check-index-only`) to fetch a few indices and then do offline queries for each concrete spec in the environment (to check if it needed to be rebuilt or not).  The alternative was to directly querying each mirror for the concrete specs in the environment, which takes a lot longer.

Reply:
> If this is the original idea, I'm happy with that, but it would be great if querying the local/offline cache caused a remote query on failure.

Did you mean it would be great if querying the local/offline cache *did not cause* a remote query on failure?  In other words, you wish we would not fetch the index on local query miss, and not try to fetch the spec directly in that case either?

Reply:
> you wish we would not fetch the index on local query miss,

Yes! I'm doing that in #34359 now.